extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/DefaultPersistenceUnitMetadata.java

```diff
@@ -0,0 +1,54 @@
+package io.quarkus.hibernate.orm.runtime;
+
+import java.util.*;
+import java.util.concurrent.*;
+
+import io.quarkus.hibernate.orm.*;
+
+import static java.util.Collections.*;
+import static java.util.stream.Collectors.*;
+
+public class DefaultPersistenceUnitMetadata implements PersistenceUnitMetadata {
+    private final Set<String> classNames;
+
+    // poor man's implementation of lazy loading: ConcurrentHashMap with only one entry
+    private final Integer DUMMY_KEY = 0;
+    private final Map<Integer, Set<Class<?>>> resolvedClasses = new ConcurrentHashMap<>();
+
+    DefaultPersistenceUnitMetadata(Set<String> classNames) {
+        this.classNames = unmodifiableSet(classNames);
+    }
+
+    @Override
+    public Set<String> getEntityClassNames() {
+        return classNames;
+    }
+
+    @Override
+    public Set<Class<?>> resolveEntityClasses(ClassLoader classLoader) {
+        return resolvedClasses.computeIfAbsent(DUMMY_KEY, ignored -> buildPersistenceUnitMetadata(classNames, classLoader));
+    }
+
+    @Override
+    public Set<Class<?>> resolveEntityClasses() {
+        return this.resolveEntityClasses(Thread.currentThread().getContextClassLoader());
+    }
+
+    private Set<Class<?>> buildPersistenceUnitMetadata(Set<String> classNames, ClassLoader classLoader) {
+        Set<Class<?>> classes = classNames.stream()
+                .map(className -> classForName(className, classLoader))
+                .collect(toSet());
```

Comment:
In Quarkus we generally try to not use lambda's in code that gets executed at runtime unless there is a very good reason. In this case a simple for-loop would do the trick, so please change the code accordingly :)

Reply:
> try to not use lambda's in code that gets executed at runtime
Just trying to understand: why?

Reply:
Because they have more runtime overhead (in terms of memory). It's not like this instance specifically will hurt memory at all, it's just that a ton of lambdas all over the place eventually will.